### PR TITLE
4.16 kvm storage issues

### PR DESCRIPTION
**Pull request by ravening**
### Description

This PR provides multiple options to handle storage issue on kvm.

Ported from https://github.com/apache/cloudstack/pull/4708

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by DaanHoogland** [comment]
@ravening sounds like this could go on 4.16. Would you?

**Comment by ravening** [comment]
> @ravening sounds like this could go on 4.16. Would you?

@DaanHoogland it's a new feature. So it will goto main

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java [hunk 1]
```diff
@@ -1311,6 +1316,26 @@ public boolean configureHostParams(final Map<String, String> params) {
...
+        }
+        if (params.get(KVM_HEARTBEAT_UPDATE_RETRY_SLEEP) != null) {
+            heartBeatUpdateRetrySleep = Long.parseLong(params.get(KVM_HEARTBEAT_UPDATE_RETRY_SLEEP));
+        }
+        if (params.get(KVM_HEARTBEAT_UPDATE_TIMEOUT) != null) {
+            heartBeatUpdateTimeout = Long.parseLong(params.get(KVM_HEARTBEAT_UPDATE_TIMEOUT));
+        }
+        if (params.get(KVM_HEARTBEAT_FAILURE_ACTION) != null) {
+            heartBeatFailureAction = KVMHAMonitor.HeartBeatAction.valueOf(params.get(KVM_HEARTBEAT_FAILURE_ACTION).toUpperCase());
+        }
```
Comment (DaanHoogland) [inline plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:1335]: could we use a `switch` statement here (or `else if`)?

Comment (DaanHoogland) [reply plugins/hypervisors/kvm/src/main/java/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:1335]: never mind, not fully awake

**Review by DaanHoogland — APPROVED** [review]
code looks good, but needs testing

**Comment by sureshanaparti** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -459,6 +463,20 @@
...
+            true, ConfigKey.Scope.Global);
+
+    public static final ConfigKey<Long> KVM_HEARTBEAT_UPDATE_RETRY_SLEEP_CK =  new ConfigKey<>("Advanced", Long.class, KVM_HEARTBEAT_UPDATE_RETRY_SLEEP, "10000",
+            "The sleep time, in milliseconds, between two kvm heartbeats to write to storage",
+            true, ConfigKey.Scope.Global);
+    public static final ConfigKey<Long> KVM_HEARTBEAT_UPDATE_TIMEOUT_CK = new ConfigKey<>("Advanced", Long.class, KVM_HEARTBEAT_UPDATE_TIMEOUT, "60000",
+            "Timeout(in milliseconds) that kvm heartbeat to write to storage",
+            true, ConfigKey.Scope.Global);
+    public static final ConfigKey<String> KVM_HEARTBEAT_FAILURE_ACTION_CK = new ConfigKey<>("Advanced", String.class, KVM_HEARTBEAT_FAILURE_ACTION, "hardreset",
+            "The action for heartbeat write failures on KVM host. The valid value are 'hardreset' (default), 'stopagent', 'destroyvms'",
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/configuration/ConfigurationManagerImpl.java:477]: The value `noaction` should be explicit here too.

File: agent/src/main/java/com/cloud/agent/properties/AgentProperties.java [hunk 3]
```diff
@@ -39,14 +39,6 @@
      */
     public static final Property<Integer> VM_MIGRATE_DOMAIN_RETRIEVE_TIMEOUT = new Property<Integer>("vm.migrate.domain.retrieve.timeout", 10);
 
-    /**
-     * Reboot host and alert management on heartbeat timeout. <br>
-     * Data type: boolean.<br>
-     * Default value: true.
-     */
-    public static final Property<Boolean> REBOOT_HOST_AND_ALERT_MANAGEMENT_ON_HEARTBEAT_TIMEOUT
-        = new Property<Boolean>("reboot.host.and.alert.management.on.heartbeat.timeout", true);
```
Comment (GutoVeronezi) [inline agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:48]: If we gonna remove this property, we need to explicit the option in the new feature that behave the same way and advise operators on how to configure it.

Also, we should remove from `agent.properties` too and wherever it is used.

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 4]
```diff
@@ -1840,4 +1878,91 @@ public void propagateChangeToAgents(Map<String, String> params) {
...
         }
     }
+
+    protected class ScanDisconnectedHostsTask extends ManagedContextRunnable {
+
+        @Override
+        protected void runInContext() {
+            try {
+                ManagementServerHostVO msHost = _msHostDao.findOneInUpState(new Filter(ManagementServerHostVO.class, "id", true, 0L, 1L));
+                if (msHost == null || (msHost.getMsid() != _nodeId)) {
```
Comment (GutoVeronezi) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:1888]:
```suggestion
                if (msHost == null || msHost.getMsid() != _nodeId) {
```

File: engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java [hunk 5]
```diff
@@ -1840,4 +1878,91 @@ public void propagateChangeToAgents(Map<String, String> params) {
...
+                        cancelInvestigationTask(hostId);
+                    }
+                    return;
+                }
+                for (HostVO host : _hostDao.listByType(Host.Type.Routing)) {
+                    if (host.getStatus() == Status.Disconnected) {
+                        scheduleInvestigationTask(host.getId());
+                    }
+                }
+            } catch (final Exception e) {
```
Comment (GutoVeronezi) [inline engine/orchestration/src/main/java/com/cloud/agent/manager/AgentManagerImpl.java:1900]: Do we need a catch Pokémon (`catch them all`) here?

File: agent/src/main/java/com/cloud/agent/properties/AgentProperties.java (hunk 3 above)
Comment (ravening) [reply agent/src/main/java/com/cloud/agent/properties/AgentProperties.java:48]: The default action is to reboot the system using the -c flag. so i will just remove it from agent.properties file

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@ravening can you look at @GutoVeronezi 's comments? I think we are about done here.

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@ravening can you look at the conflicts, please?

**Comment by DaanHoogland** [comment]
@ravening is this still relevant?

**Pull request closed without merging**